packages/framer-motion/src/render/types.ts

```diff
@@ -170,14 +37,109 @@ export type VisualElementOptions<Instance, RenderState = any> = {
     [key: string]: string | number
 }
+
+export interface VisualElementEventCallbacks {
+    BeforeLayoutMeasure: () => void
+    LayoutMeasure: (layout: Box, prevLayout?: Box) => void
+    LayoutUpdate: (layout: Axis, prevLayout: Axis) => void
+    Update: (latest: ResolvedValues) => void
+    Render: () => void
+    AnimationStart: (definition: AnimationDefinition) => void
+    AnimationComplete: (definition: AnimationDefinition) => void
+    LayoutAnimationStart: () => void
+    LayoutAnimationComplete: () => void
+    SetAxisTarget: () => void
+    Unmount: () => void
+}
+
+export interface LayoutLifecycles {
+    onBeforeLayoutMeasure?(box: Box): void
+
+    onLayoutMeasure?(box: Box, prevBox: Box): void
+
+    /**
+     * @internal
+     */
+    onLayoutAnimationStart?(): void
+
+    /**
+     * @internal
+     */
+    onLayoutAnimationComplete?(): void
+}
+
+export interface AnimationLifecycles {
+    /**
+     * Callback with latest motion values, fired max once per frame.
+     *
+     * ```jsx
+     * function onUpdate(latest) {
+     *   console.log(latest.x, latest.opacity)
+     * }
+     *
+     * <motion.div animate={{ x: 100, opacity: 0 }} onUpdate={onUpdate} />
+     * ```
+     */
+    onUpdate?(latest: ResolvedValues): void
+
+    /**
+     * Callback when animation defined in `animate` begins.
+     *
+     * The provided callback will be called with the triggering animation definition.
+     * If this is a variant, it'll be the variant name, and if a target object
+     * then it'll be the target object.
+     *
+     * This way, it's possible to figure out which animation has started.
+     *
+     * ```jsx
+     * function onStart() {
+     *   console.log("Animation started")
+     * }
+     *
+     * <motion.div animate={{ x: 100 }} onAnimationStart={onStart} />
+     * ```
+     */
+    onAnimationStart?(definition: AnimationDefinition): void
+
+    /**
+     * Callback when animation defined in `animate` is complete.
+     *
+     * The provided callback will be called with the triggering animation definition.
+     * If this is a variant, it'll be the variant name, and if a target object
+     * then it'll be the target object.
+     *
+     * This way, it's possible to figure out which animation has completed.
+     *
+     * ```jsx
+     * function onComplete() {
+     *   console.log("Animation completed")
+     * }
+     *
+     * <motion.div
+     *   animate={{ x: 100 }}
+     *   onAnimationComplete={definition => {
+     *     console.log('Completed animating', definition)
+     *   }}
+     * />
+     * ```
+     */
+    onAnimationComplete?(definition: AnimationDefinition): void
+
+    /**
+     * @internal
+     */
+    onUnmount?(): void
+}
+
+export type EventProps = LayoutLifecycles & AnimationLifecycles
+
+export type CreateVisualElement<Instance> = (
+    Component: string | React.ComponentType<React.PropsWithChildren<unknown>>,
+    options: VisualElementOptions<Instance>
+) => VisualElement<Instance>
```

Comment:
Why is this type needed? We shoud be able to use the constructor form VisualElement, right?

Reply:
This is a function that returns **a** visual element. For our DOM motion components this can be either `HTMLVisualElement` or `SVGVisualElement`

Reply:
Can we make it specific to those visual elements then?
Don't feel strongly, because I don't have full context, so feel free to leave as is too.